File: Sources/PauselyCore/TimerTiming.swift
import Foundation

public enum TimerTiming {
    /// The transition happens on the tick that reaches zero, not one tick later.
    public static func countdown(_ remaining: Int) -> Int { max(0, remaining - 1) }

    public struct Position {
        public let cycle: Int
        public let isBreak: Bool
        public let remaining: Int
    }

    public static func position(work: Double, rest: Double, anchor: Double, now: Double) -> Position {
        let duration = work + rest
        let elapsed = max(0, now - anchor)
        let position = elapsed.truncatingRemainder(dividingBy: duration)
        return Position(cycle: Int(elapsed / duration), isBreak: position >= work,
                        remaining: Int(ceil(position < work ? work - position : duration - position)))
    }
}

File: PauselyWindows/Services/TimerTiming.cs
namespace PauselyWindows.Services;

internal static class TimerTiming
{
    // Transition on the tick that reaches zero, not one tick later.
    internal static int Countdown(int remaining) => Math.Max(0, remaining - 1);

    internal readonly record struct Position(long Cycle, bool IsBreak, int Remaining);

    internal static Position At(double work, double rest, double anchor, double now)
    {
        double duration = work + rest;
        double elapsed = Math.Max(0, now - anchor);
        double position = elapsed % duration;
        return new((long)(elapsed / duration), position >= work,
            (int)Math.Ceiling(position < work ? work - position : duration - position));
    }
}
